### PR TITLE
Automated QA Pipeline for Android App (and Other CI Tweaks)

### PR DESCRIPTION
**Pull request by MarkDavies0**
## Changes
Adds steps to the `build` action to build the android app and type check the mobile directory. Also splits with existing checks into smaller workflows.

## Requirements
N/A

## Subtasks / Progress
- [x] Add build job for Android
- ~~[ ] Add build job for iOS (currently blocked by GitHub Actions' lack of support for macOS-11 and Xcode 12.5)~~
- [x] Add lint job for mobile directory (errors to be fixed in #85)

## Testing

Check the build action to see logs of the android build

- [x] Automated

List of added tests:
- install mobile
- build android
- lint mobile

### CODE REVIEW

**Comment by rlnsy** [comment]
@MarkDavies0 would it be possible to just do a simple tsc pass based on the errors you fixed in that one branch? Or do we need cocoa pods etc. in order to do that.

**Comment by MarkDavies0** [comment]
@rlnsy I have added a tsc pass in mobile (no cocoapods required) however there are a bunch of errors caused by importing Firebase modules. I'm wondering if you have some insight on these errors.

**Comment by rlnsy** [comment]
@MarkDavies0 So after messing around with the builds for a while, I reapplied (or rather, remembered) that there is a strange element to the dependency structure of this project which is that the common subdirectory has no package.json/node_modules of it's own. Since the files there are used by both the server and mobile projects, the dependency handling is done on a per-platform basis, and in the case of the mobile apps certain dependencies like firebase are actually linked from the pods/gradle. This means that we cannot use regular tsc on those codebases and expect it to build without error. To be more concrete, take a look at the kind of errors we get when doing tsc right now:
```
../common/database/repositories/dbProvider.ts(2,32): error TS2307: Cannot find module 'firebase-admin' or its corresponding type declarations.
```
Sure enough, if you look at the mobile/package.json, there is no firebase-admin listed there. The android build works great because it actually uses the gradle build system which is supported in the GitHub environment (the iOS equivalent is the now-in-preview Xcode support in GitHub as you mentioned).

I propose we adopt one of the following two solutions (open to discussion on which one you think is best - @wynnset  maybe you want to chime in here):
1. Get rid of the tsc step, limit the scope of this PR to setting up the Android build (which is already done), and wait for the GH Xcode/Mac toolchain support before opening up an addition PR to incorporate the iOS pipeline.
2. Also wait for the iOS support to set up an iOS-specific pipeline - but instead of getting rid of the tsc step, install the missing modules to the mobile directory to get tsc to pass (I'm not a huge fan of this as there may be issues and we'll have to manage two different versions/references to the dependencies, which could make our build less representative of the state of the iOS codebase)

In either case, we should also add a lint step for the entire mobile directory.

**Comment by rlnsy** [comment]
For reference: here are the results with the firebase dependency installed: https://github.com/bipolarbridges/companion-app/actions/runs/1009702078. There are still some errors coming from the persona repo I think.

**Comment by wynnset** [comment]
> 1. Get rid of the tsc step, limit the scope of this PR to setting up the Android build (which is already done), and wait for the GH Xcode/Mac toolchain support before opening up an addition PR to incorporate the iOS pipeline.

I like this solution

> In either case, we should also add a lint step for the entire mobile directory.

Yes, agreed. @rlnsy we discussed on Tuesday linting the mobile directory as well. One caveat is that it might cause merge conflicts when we want to merge in changes from Maslo. However, we can reduce those merge conflicts if we lint their code before we merge it into our code. I suspect it will still cause some merge conflicts even with that method, but it won't be as bad. Perhaps we should also ask Maslo to lint their mobile directory (or find out why they haven't) - that would be the ideal solution.

**Comment by MarkDavies0** [comment]
> Sure enough, if you look at the mobile/package.json, there is no firebase-admin listed there. The android build works great because it actually uses the gradle build system which is supported in the GitHub environment (the iOS equivalent is the now-in-preview Xcode support in GitHub as you mentioned).
> 

@rlnsy Thanks for the help Rowan this makes a lot of sense. Do you think there could be a way to silence these firebase-errors as they were the only ones holding me up? (Persona and such can be fixed) I would really like to include the TSC step but it may have to been done on the developers side.

> 1. Get rid of the tsc step, limit the scope of this PR to setting up the Android build (which is already done), and wait for the GH Xcode/Mac toolchain support before opening up an addition PR to incorporate the iOS pipeline.

This seems like the best option for now

>In either case, we should also add a lint step for the entire mobile directory. 

+1

**Comment by rlnsy** [comment]
@MarkDavies0 can you try and remove the lint-related changes? Since we are listing in masco upstream first.

**Comment by MarkDavies0** [comment]
@rlnsy should be good to go now

File: mobile/src/hooks/getHealthKitData.tsx
Comment (rlnsy) [inline mobile/src/hooks/getHealthKitData.tsx:36]: What's going on here? Should we try to get it to work rather than commenting out?

Comment (MarkDavies0) [reply mobile/src/hooks/getHealthKitData.tsx:36]: I think it was throwing TSC errors and I knew @SukiCai was working on this so I just commented it out. Probably best to revert to master state

**Comment by rlnsy** [comment]
@MarkDavies0 this looks like it had some merged branches and needs to be rebased. Do you remember what the old base was?

**Comment by MarkDavies0** [comment]
@rlnsy Does this need to be rebased as it can be merged directly into master?

**Comment by rlnsy** [comment]
@MarkDavies0 no you're right, it doesn't and we can squash these changes into a single commit. I was hoping to rebase something else on top of this but that proved too hard with all the commits from the other branches, so I'll hold off until this is merged.

**Review by wynnset — APPROVED** [review]

**Comment by wynnset** [comment]
Code looks good. Just need to fix the merge conflicts and the failing test

**Comment by rlnsy** [comment]
@wynnset the failing test will not be fixed as it is the mobile lint which you didn't want us to fix before doing it in the masco repo.

**Comment by MarkDavies0** [comment]
@rlnsy

` Linking CompanionKit
⚠️  ld: Could not find or use auto-linked library 'swiftUniformTypeIdentifiers'

❌  ld: symbol(s) not found for architecture x86_64



❌  clang: error: linker command failed with exit code 1 (use -v to see invocation)`

**Pull request merged**
